### PR TITLE
Port to python3, dockerize and add some args

### PR DESCRIPTION
**Pull request by simeonmiteff**
Hi @thepacketgeek 

Thanks for this tool. I've:
- Made it work with @ofalk's (fork?) of libdnet: http://github.com/ofalk/libdnet
- Converted to Python 3
- Added a Dockerfile
- Added an option to overwrite instead of append to output files (now default)`*`
- Added a VLAN tag rewriting feature (fixed VLAN only, no support for multiple tags and no randomisation)

`*` not sure if this is wise in general but is _my_ preference.

### CODE REVIEW

**Comment by simeonmiteff** [comment]
Oh I forgot to add - I haven't changed the documentation (README.md) - I'd like to hear where you'd like to go with this PR before reworking it.

**Comment by thepacketgeek** [comment]
Thank you for bring a fresh breath of air to this project :) I'll look through the commits and let you know, I think all of these are wonderful changes bringing sanicap up to date!

File: sanicap/sanicap.py [hunk 1]
```diff
@@ -51,7 +53,7 @@ def _random_mac(self, address):
         def pad_bin(unpadded):
             return format(int('0x' + unpadded.replace(':','').replace('.',''), 16), '048b')
 
-        unmasked = ''.join([str(randint(0,1)) for x in xrange(0, 48 - self.mask)])
+        unmasked = ''.join([str(randint(0,1)) for x in range(0, 48 - self.mask)])
```
Comment (thepacketgeek) [inline sanicap/sanicap.py]: I realize this was existing: join works with an iterator, no need to create an intermediate list
```suggestion
        unmasked = ''.join(str(randint(0,1)) for x in range(0, 48 - self.mask))
```

File: sanicap/sanicap.py [hunk 2]
```diff
@@ -113,9 +115,9 @@ def pad_bin(unpadded):
     def _random_ip(self, address):
 
         def pad_bin(unpadded):
-            return format(int(ipaddress.IPv4Address(unicode(unpadded))), '032b')
+            return format(int(ipaddress.IPv4Address(str(unpadded))), '032b')
 
-        unmasked = ''.join([str(randint(0,1)) for x in xrange(0, 32 - self.mask)])
+        unmasked = ''.join([str(randint(0,1)) for x in range(0, 32 - self.mask)])
```
Comment (thepacketgeek) [inline sanicap/sanicap.py]: Same as above
```suggestion
        unmasked = ''.join(str(randint(0,1)) for x in range(0, 32 - self.mask))
```

File: sanicap/sanicap.py [hunk 3]
```diff
@@ -181,9 +183,9 @@ def pad_bin(unpadded):
     def _random_ip(self, address):
 
         def pad_bin(unpadded):
-            return format(int(ipaddress.IPv6Address(unicode(unpadded))), '0128b')
+            return format(int(ipaddress.IPv6Address(str(unpadded))), '0128b')
 
-        unmasked = ''.join([str(randint(0,1)) for x in xrange(0, 128 - self.mask)])
+        unmasked = ''.join([str(randint(0,1)) for x in range(0, 128 - self.mask)])
```
Comment (thepacketgeek) [inline sanicap/sanicap.py]: Same as above
```suggestion
        unmasked = ''.join(str(randint(0,1)) for x in range(0, 128 - self.mask))
```

File: sanicap/sanicap.py [hunk 4]
```diff
@@ -217,7 +219,7 @@ def get_ip(self, address):
             return address
 
 
-def sanitize(filepath_in, filepath_out = None, sequential=True, ipv4_mask=0, ipv6_mask=0, mac_mask=0, start_ipv4='10.0.0.1', start_ipv6='2001:aa::1', start_mac='00:aa:00:00:00:01', info=True):
+def sanitize(filepath_in, filepath_out = None, sequential=True, append=False, ipv4_mask=0, ipv6_mask=0, mac_mask=0, start_ipv4='10.0.0.1', start_ipv6='2001:aa::1', start_mac='00:aa:00:00:00:01', fixed_vlan=None, info=True):
```
Comment (thepacketgeek) [inline sanicap/sanicap.py]: In lieu of an `Options` or `Config` class for all of these options, making them kwarg only should help usage stay readable.

```suggestion
def sanitize(filepath_in, *, filepath_out = None, sequential=True, append=False, ipv4_mask=0, ipv6_mask=0, mac_mask=0, start_ipv4='10.0.0.1', start_ipv6='2001:aa::1', start_mac='00:aa:00:00:00:01', fixed_vlan=None, info=True):
```

**Review by thepacketgeek — CHANGES_REQUESTED** [review]
Love it! Just a few small changes, thank you for this

**Review by thepacketgeek — APPROVED** [review]

**Comment by thepacketgeek** [comment]
This is great, thanks again! Let me know if you are still planning on README changes or if this is ready to merge

**Comment by simeonmiteff** [comment]
@thepacketgeek I just realised sanicap doesn't need `dnet` :facepalm:

**Comment by simeonmiteff** [comment]
> This is great, thanks again! Let me know if you are still planning on README changes or if this is ready to merge

I reworked the README and while generating an example output I discovered that the kwarg change introduced a bug, so I fixed that and amended that commit. I also fixed/improved the error handling and fixed some formatting things that PyCharm complained about.

I also revised the Dockerfile commit and ripped all the `libdnet` building garbage.

The Dockerfile is specifically useful to be because I'm running sanicap on hosts without web access so just scp-ing the image over is easier than sorting out offline installs of the 3 dependencies. I'd like to keep it in the repo but I recon this module really also needs proper python packaging. I'm reaching the limit of how much time I can justify hacking on this now.

TBH I'd love to write something like sanicap in Rust (not just for RIIR sake: sanicap is _slow_ on a 1.5GB PCAP). I have some good code for handling both pcap and pcapng inputs but scapy does a huge amount of heavy lifting dealing with the link layer all the way up the stack - that would be hard to replicate in Rust (unless there are some awesome crates I have yet to discover).

**Comment by thepacketgeek** [comment]
The Dockerfile does seem very useful good to keep in the repo. 

I agree that this project could benefit greatly from a rewrite in Rust, it's something that I was pondering when I realized there was no scapy equivalent in Rust. I started on some of the underlying features for the send/receive packet flows in [`rust-network-tools`](https://github.com/thepacketgeek/rust-network-tools), but modifying and crafting packets has been a pain in the libraries I've tried so far ([`etherparse`](https://crates.io/crates/resynth) and [`pnet`](https://crates.io/crates/pnet)). [`Resynth`](https://crates.io/crates/resynth) looks promising but it'll be a while before anything catches up to Scapy with protocol support.

**Comment by ofalk** [comment]
[ ... ]
> Thanks for this tool. I've:
>     * Made it work with @ofalk's (fork?) of libdnet: [http://github.com/ofalk/libdnet](https://github.com/ofalk/libdnet)

It used to be a fork, but quite a while ago I took over the project from it's original author and now it's the official one, just FYI.

**Comment by ofalk** [comment]
> @thepacketgeek I just realised sanicap doesn't need `dnet` facepalm

Great, I'm out then :-)

**Pull request merged**